pkg/agent/cniserver/ipam/antrea_ipam_controller.go

```diff
@@ -95,16 +70,42 @@ func podIndexFunc(obj interface{}) ([]string, error) {
 	return podNames.UnsortedList(), nil
 }
 
-func InitializeAntreaIPAMController(kubeClient clientset.Interface, crdClient clientsetversioned.Interface, informerFactory informers.SharedInformerFactory, podInformer cache.SharedIndexInformer, crdInformerFactory externalversions.SharedInformerFactory) (*AntreaIPAMController, error) {
-	antreaIPAMController := NewAntreaIPAMController(kubeClient, crdClient, informerFactory, podInformer, crdInformerFactory)
-
+func InitializeAntreaIPAMController(crdClient clientsetversioned.Interface,
+	informerFactory informers.SharedInformerFactory,
+	crdInformerFactory externalversions.SharedInformerFactory,
+	podInformer cache.SharedIndexInformer, ipamAnnotations bool) (*AntreaIPAMController, error) {
 	// Order of init causes antreaIPAMDriver to be initialized first
 	// After controller is initialized by agent init, we need to make it
 	// know to the driver
 	if antreaIPAMDriver == nil {
 		return nil, fmt.Errorf("Antrea IPAM driver failed to initialize")
 	}
 
+	var antreaIPAMController *AntreaIPAMController
+	ipPoolInformer := crdInformerFactory.Crd().V1alpha2().IPPools()
+	ipPoolInformer.Informer().AddIndexers(cache.Indexers{podIndex: podIndexFunc})
+
+	// Create podInformer/Lister and namespaceInformer/Lister if need to read the AntreaIPAM
+	// annotation on Pods and Namespaces.
+	if ipamAnnotations {
+		namespaceInformer := informerFactory.Core().V1().Namespaces()
+		antreaIPAMController = &AntreaIPAMController{
+			crdClient:         crdClient,
+			ipPoolInformer:    ipPoolInformer,
+			ipPoolLister:      ipPoolInformer.Lister(),
+			namespaceInformer: namespaceInformer,
+			namespaceLister:   namespaceInformer.Lister(),
+			podInformer:       podInformer,
+			podLister:         corelisters.NewPodLister(podInformer.GetIndexer()),
+		}
+	} else {
+		antreaIPAMController = &AntreaIPAMController{
+			crdClient:      crdClient,
+			ipPoolInformer: ipPoolInformer,
+			ipPoolLister:   ipPoolInformer.Lister(),
+		}
+	}
+	antreaIPAMDriver.setController(antreaIPAMController)
```

Comment:
antreaIPAMDriver also calls setController in line129, why do we need twice?

Reply:
It must be my mistake. Thanks!